### ISSUE TITLE
💅 noUndeclaredVariables does not raise an error for variables used before initialisation

### ISSUE THREAD

**Issue by th1nkful** [biomejs/biome#2643]
### Environment information

```bash
CLI:
  Version:                      1.6.1
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.11.1"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "pnpm/9.0.6"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Linter:
  Recommended:                  true
  All:                          false
  Rules:                        a11y/all = false
                                complexity/noBannedTypes = "off"
                                complexity/noForEach = "off"
                                complexity/noStaticOnlyClass = "off"
                                complexity/noUselessThisAlias = "off"
                                complexity/useFlatMap = "off"
                                complexity/useOptionalChain = "off"
                                correctness/noUndeclaredVariables = "error"
                                correctness/noUnusedImports = "error"
                                correctness/useExhaustiveDependencies = "off"
                                performance/noAccumulatingSpread = "off"
                                performance/noDelete = "off"
                                security/noDangerouslySetInnerHtml = "off"
                                style/noInferrableTypes = "off"
                                style/noNonNullAssertion = "off"
                                style/noParameterAssign = "off"
                                style/useEnumInitializers = "off"
                                style/useNodejsImportProtocol = "off"
                                style/useNumberNamespace = "off"
                                suspicious/noConfusingVoidType = "off"
                                suspicious/noConsoleLog = "error"
                                suspicious/noEmptyInterface = "off"
                                suspicious/noExplicitAny = "off"
                                suspicious/noShadowRestrictedNames = "off"
                                suspicious/noThenProperty = "off"

Workspace:
  Open Documents:               0
```

### Rule name

noUndeclaredVariables

### Playground link

https://biomejs.dev/playground/?lintRules=all&code=aQBtAHAAbwByAHQAIABQAHIAbwBtAGkAcwBlAHMAIABmAHIAbwBtACAAJwBiAGwAdQBlAGIAaQByAGQAJwA7AAoACgBjAG8AbgBzAHQAIABnAGUAdABQAGEAZwBlAGQAQwBvAG4AdABlAG4AdABzACAAPQAgAGEAcwB5AG4AYwAgACgAYwB1AHMAdABvAG0AZQByAEkAZAAsACAAbwBwAHQAaQBvAG4AcwApACAAPQA%2BACAAewAKACAAIAAvAC8AIABoAGkAdAAgAEQAQgAKACAAIABhAHcAYQBpAHQAIABuAGUAdwAgAFAAcgBvAG0AaQBzAGUAKAAoAHIAZQBzAG8AbAB2AGUAKQAgAD0APgAgAHMAZQB0AFQAaQBtAGUAbwB1AHQAKAByAGUAcwBvAGwAdgBlACwAIAAyADUAMAApACkAOwAKAAoAIAAgAHIAZQB0AHUAcgBuACAAWwB7AAoAIAAgACAAIABjAHUAcwB0AG8AbQBlAHIASQBkACwACgAgACAAIAAgAC4ALgAuAG8AcAB0AGkAbwBuAHMALAAKACAAIAB9AF0AOwAKAH0AOwAKAAoAYwBvAG4AcwB0ACAAZgBvAHIAbQBhAHQAUwB1AG0AbQBhAHIAeQBJAHQAZQBtACAAPQAgACgAYQBjAHQAaQBvAG4ALAAgAGMAbwBuAHQAZQBuAHQALAAgAG8AcAB0AGkAbwBuAHMAKQAgAD0APgAgAHsACgAgACAAcgBlAHQAdQByAG4AIAB7AAoAIAAgACAAIAAuAC4ALgBhAGMAdABpAG8AbgAsAAoAIAAgACAAIAAuAC4ALgBjAG8AbgB0AGUAbgB0ACwACgAgACAAIAAgAC4ALgAuAG8AcAB0AGkAbwBuAHMALAAKACAAIAB9ADsACgB9AAoACgBjAG8AbgBzAHQAIABlAHgAYQBtAHAAbABlACAAPQAgAGEAcwB5AG4AYwAgACgAYwB1AHMAdABvAG0AZQByACwAIABhAGMAdABpAG8AbgAsACAAYQBjAHQAaQBvAG4ASQB0AGUAbQBzAEcAcgBvAHUAcABlAGQAKQAgAD0APgAgAHsACgAgACAAYwBvAG4AcwB0ACAAYwBvAG4AdABlAG4AdABzADoAIABSAGUAYwBvAHIAZAA8AHMAdAByAGkAbgBnACwAIABzAHQAcgBpAG4AZwA%2BAFsAXQAgAD0AIABbAF0AOwAKAAoAIAAgAGEAdwBhAGkAdAAgAFAAcgBvAG0AaQBzAGUAcwAuAHAAYQByAGEAbABsAGUAbAAoAGEAYwB0AGkAbwBuAEkAdABlAG0AcwBHAHIAbwB1AHAAZQBkACwAIABhAHMAeQBuAGMAIAAoAGkAdABlAG0AcwApACAAPQA%2BACAAewAKACAAIAAgACAAYwBvAG4AcwB0ACAAYwBvAG4AdABlAG4AdABJAHQAZQBtAHMAIAA9ACAAYQB3AGEAaQB0ACAAZwBlAHQAUABhAGcAZQBkAEMAbwBuAHQAZQBuAHQAcwAoAGMAdQBzAHQAbwBtAGUAcgAuAGkAZAAsACAAewAKACAAIAAgACAAIAAgAGkAZABzADoAIAAgAGkAdABlAG0AcwAuAG0AYQBwACgAKABpAHQAZQBtACkAIAA9AD4AIABpAHQAZQBtAC4AaQBkACkALAAKACAAIAAgACAAIAAgAHQAaQBtAGUAegBvAG4AZQAsACAALwAvACAALQA%2BACAAZQB4AHAAZQBjAHQAaQBuAGcAIABhAG4AIABlAHIAcgBvAHIAIABoAGUAcgBlACAAZgBvAHIACgAgACAAIAAgAH0AKQA7AAoACgAgACAAIAAgAGMAbwBuAHQAZQBuAHQAcwAuAHAAdQBzAGgAKAAuAC4ALgBjAG8AbgB0AGUAbgB0AEkAdABlAG0AcwApADsACgAgACAAfQApADsACgAKACAAIABjAG8AbgBzAHQAIAB0AGkAbQBlAHoAbwBuAGUAIAA9ACAAYQBjAHQAaQBvAG4ALgB0AGkAbQBlAHoAbwBuAGUAOwAKACAAIAAKACAAIABjAG8AbgBzAHQAIABpAHQAZQBtAHMAIAA9ACAAYwBvAG4AdABlAG4AdABzAC4AbQBhAHAAKAAoAGMAbwBuAHQAZQBuAHQAKQAgAD0APgAKACAAIAAgACAAZgBvAHIAbQBhAHQAUwB1AG0AbQBhAHIAeQBJAHQAZQBtACgAYQBjAHQAaQBvAG4ALAAgAGMAbwBuAHQAZQBuAHQALAAgAHsAIAB0AGkAbQBlAHoAbwBuAGUAIAB9ACkALAAKACAAIAApADsACgAKACAAIAByAGUAdAB1AHIAbgAgAGkAdABlAG0AcwA7AAoAfQAKAAoAZQB4AGEAbQBwAGwAZQAoACkAOwA%3D

### Expected result

Expected biome to error for variable

Im the repro, I am using a variable (timezone) before initialisation (error I saw at runtime was "ReferenceError: Cannot access 'timezone' before initialization"). Maybe I am misunderstanding how the rule works, but I do remember that eslint would mark these kinds of mistakes

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by Conaclos** [comment]
This is a particularly hard case to catch. [TSC](https://www.typescriptlang.org/play/?noImplicitAny=false#code/JYWwDg9gTgLgBABShEwDOBTNcBmyRwDkARgDYCuGxwUAJoQNwBQTAxhAHZrwDmGMCAIZ9aAYU4wMHGNgC8cQWgCeHVnAAUrctxQYoASVoAaOBDAxgnNAEo4sgHxwA3kzhwA9O7gALYPAAiAEKuCgDugn5wHBihiPjoGOrqUFgQpABuGLYOcJgwACqgGBDkMMmpGRgmAEwArAAM1tbMISkw5FAccADaLm5uWjogeoZGIW4AdFNmFlZjbgC+ALrMCy3sXPA40CCCMADK5CC7UEr6kgTy6oKssxwmG5LSJjOWXNmOfXBtHV1fk1Mbnd5v0phNHlIYCCARNXnMQmsmAsWBtuHAMAAPQTgUgYOwKZSqDSDGC6KAmIFvCm3N7nDAgNAAcWQ5DAGFoH2cIVR8Ah0jQAC44AAlDDsOgAHm4UGAHB4Jmlsp49m6S3xqpabkE4UiSBQCTQEzAgigglIuNI1xpnDpDOZJTZxgJKjU6j89Jsdk+4zgPN9EkhtrkYQivH4QhE4mkkLQmm0pOGUAmwCd-36KcFbndDImuzASWznOzyY50P6FmGAC9OFUPF4ALSOTFsmlyhRdPTIKA+PR47ZQH0LZosfr+6P8o3abzqMF8mBB4eLYfcqzwCsYavRfGUzgTdebjDMNwrzZwbPBueGvNJOcfH393YHI4nM4XK3AsdPKHOOD7mtwIcQWXNwfk6M8LjQVYWExbEwFxdRmiAA) and [ESLint](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW5kZWY6IFwiZXJyb3JcIiovXG5cbmltcG9ydCBQcm9taXNlcyBmcm9tICdibHVlYmlyZCc7XG5cbmNvbnN0IGdldFBhZ2VkQ29udGVudHMgPSBhc3luYyAoY3VzdG9tZXJJZCwgb3B0aW9ucykgPT4ge1xuICAvLyBoaXQgREJcbiAgYXdhaXQgbmV3IFByb21pc2UoKHJlc29sdmUpID0+IG51bGwpO1xuXG4gIHJldHVybiBbe1xuICAgIGN1c3RvbWVySWQsXG4gICAgLi4ub3B0aW9ucyxcbiAgfV07XG59O1xuXG5jb25zdCBmb3JtYXRTdW1tYXJ5SXRlbSA9IChhY3Rpb24sIGNvbnRlbnQsIG9wdGlvbnMpID0+IHtcbiAgcmV0dXJuIHtcbiAgICAuLi5hY3Rpb24sXG4gICAgLi4uY29udGVudCxcbiAgICAuLi5vcHRpb25zLFxuICB9O1xufVxuXG5jb25zdCBleGFtcGxlID0gYXN5bmMgKGN1c3RvbWVyLCBhY3Rpb24sIGFjdGlvbkl0ZW1zR3JvdXBlZCkgPT4ge1xuICBjb25zdCBjb250ZW50cyA9IFtdO1xuXG4gIGF3YWl0IFByb21pc2VzLnBhcmFsbGVsKGFjdGlvbkl0ZW1zR3JvdXBlZCwgYXN5bmMgKGl0ZW1zKSA9PiB7XG4gICAgY29uc3QgY29udGVudEl0ZW1zID0gYXdhaXQgZ2V0UGFnZWRDb250ZW50cyhjdXN0b21lci5pZCwge1xuICAgICAgaWRzOiAgaXRlbXMubWFwKChpdGVtKSA9PiBpdGVtLmlkKSxcbiAgICAgIHRpbWV6b25lLCAvLyAtPiBleHBlY3RpbmcgYW4gZXJyb3IgaGVyZSBmb3JcbiAgICB9KTtcblxuICAgIGNvbnRlbnRzLnB1c2goLi4uY29udGVudEl0ZW1zKTtcbiAgfSk7XG5cbiAgY29uc3QgdGltZXpvbmUgPSBhY3Rpb24udGltZXpvbmU7XG4gIFxuICBjb25zdCBpdGVtcyA9IGNvbnRlbnRzLm1hcCgoY29udGVudCkgPT5cbiAgICBmb3JtYXRTdW1tYXJ5SXRlbShhY3Rpb24sIGNvbnRlbnQsIHsgdGltZXpvbmUgfSksXG4gICk7XG5cbiAgcmV0dXJuIGl0ZW1zO1xufVxuXG5leGFtcGxlKCk7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==) doesn't report any issue either.

The issue comes from the execution of the promise before the declaration of a variable used in the callback.

**Comment by th1nkful** [comment]
A dev on my team pointed out that it's a different [eslint rule](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW5kZWY6IFwiZXJyb3JcIiovXG5cbmltcG9ydCBQcm9taXNlcyBmcm9tICdibHVlYmlyZCc7XG5cbmNvbnN0IGdldFBhZ2VkQ29udGVudHMgPSBhc3luYyAoY3VzdG9tZXJJZCwgb3B0aW9ucykgPT4ge1xuICAvLyBoaXQgREJcbiAgYXdhaXQgbmV3IFByb21pc2UoKHJlc29sdmUpID0+IG51bGwpO1xuXG4gIHJldHVybiBbe1xuICAgIGN1c3RvbWVySWQsXG4gICAgLi4ub3B0aW9ucyxcbiAgfV07XG59O1xuXG5jb25zdCBmb3JtYXRTdW1tYXJ5SXRlbSA9IChhY3Rpb24sIGNvbnRlbnQsIG9wdGlvbnMpID0+IHtcbiAgcmV0dXJuIHtcbiAgICAuLi5hY3Rpb24sXG4gICAgLi4uY29udGVudCxcbiAgICAuLi5vcHRpb25zLFxuICB9O1xufVxuXG5jb25zdCBleGFtcGxlID0gYXN5bmMgKGN1c3RvbWVyLCBhY3Rpb24sIGFjdGlvbkl0ZW1zR3JvdXBlZCkgPT4ge1xuICBjb25zdCBjb250ZW50cyA9IFtdO1xuXG4gIGF3YWl0IFByb21pc2VzLnBhcmFsbGVsKGFjdGlvbkl0ZW1zR3JvdXBlZCwgYXN5bmMgKGl0ZW1zKSA9PiB7XG4gICAgY29uc3QgY29udGVudEl0ZW1zID0gYXdhaXQgZ2V0UGFnZWRDb250ZW50cyhjdXN0b21lci5pZCwge1xuICAgICAgaWRzOiAgaXRlbXMubWFwKChpdGVtKSA9PiBpdGVtLmlkKSxcbiAgICAgIHRpbWV6b25lLCAvLyAtPiBleHBlY3RpbmcgYW4gZXJyb3IgaGVyZSBmb3JcbiAgICB9KTtcblxuICAgIGNvbnRlbnRzLnB1c2goLi4uY29udGVudEl0ZW1zKTtcbiAgfSk7XG5cbiAgY29uc3QgdGltZXpvbmUgPSBhY3Rpb24udGltZXpvbmU7XG4gIFxuICBjb25zdCBpdGVtcyA9IGNvbnRlbnRzLm1hcCgoY29udGVudCkgPT5cbiAgICBmb3JtYXRTdW1tYXJ5SXRlbShhY3Rpb24sIGNvbnRlbnQsIHsgdGltZXpvbmUgfSksXG4gICk7XG5cbiAgcmV0dXJuIGl0ZW1zO1xufVxuXG5leGFtcGxlKCk7Iiwib3B0aW9ucyI6eyJydWxlcyI6eyJuby11c2UtYmVmb3JlLWRlZmluZSI6WyJlcnJvciJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7InNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19) that detects this - "no-use-before-define" 

Is their similar in biome yet?

**Comment by Conaclos** [comment]
> Is their similar in biome yet?

Yes it is [noInvalidUseBeforeDeclaration](https://biomejs.dev/linter/rules/no-invalid-use-before-declaration/).
The rule doesn't detect the issue.

**Comment by Conaclos** [comment]
> Is their similar in biome yet?

Yes it is [noInvalidUseBeforeDeclaration](https://biomejs.dev/linter/rules/no-invalid-use-before-declaration/).

Note that the ESLint rule reports false positives.
For instance, the following code should be valid:

```js
const f = () => x;
const x = 0;
f();
```

Conversely, the Biome rule misses some cases.
We could try to improve it.
